### PR TITLE
feat: Move additional event handler functionality to `Client`

### PR DESCRIPTION
**Pull request by davidhozic**
## Summary
The pull request introduces changes proposed in #1906 .

It allows the user to register multiple ``on_`` event handlers inside the ``Client`` objects instead of just in``Bot``. 

```py
import discord
import secret


class MyClient(discord.Client):
    async def on_ready(self):
        print("Default handler")


intents = discord.Intents.default()
intents.message_content = True
client = MyClient(intents=intents)


@client.event # Override the default handler
async def on_ready():
    print(f"Handler 1: {client.user.name} has logged in!")

@client.listen # Additional handler
async def on_ready():
    print(f"Handler 2: {client.user.name} has logged in!")

@client.listen("on_ready") # Additional handler
async def ready_handler():
    print(f"Handler 3: {client.user.name} has logged in!")

@client.listen("on_ready") # Additional handler
async def ready_handler():
    print(f"Handler 4: {client.user.name} has logged in!")

@client.listen()
async def on_message(message: discord.Message):
    print(f"Received message {message.content}")


@client.listen("on_guild_channel_update")
async def channel_updated(prev: discord.ChannelType, curr:discord.ChannelType):
    print(prev.name, curr.name)

client.run(secret.TOKEN)
```

Also add support to allow ``@listen`` without ``()`` at the end.
```py
import discord
import secret

bot = discord.Bot()

@bot.listen()
async def on_ready():
    print(f"Handler 1: {bot.user.name} has logged in!")

@bot.listen # <-------- No ()
async def on_ready():
    print(f"Handler 2: {bot.user.name} has logged in!")

bot.run(secret.TOKEN)
```

## Information

- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [x] If `type: ignore` comments were used, a comment is also left explaining why.

### CODE REVIEW

File: discord/client.py [hunk 1]
```diff
@@ -246,12 +246,18 @@ def __init__(
         self._ready: asyncio.Event = asyncio.Event()
         self._connection._get_websocket = self._get_websocket
         self._connection._get_client = lambda: self
+        self._event_handlers: dict[str, list[Coro]] = {}
+        self._prepeare_handlers()
```
Comment (VincentRPS) [inline discord/client.py:250]:
```suggestion
        self._prepare_handlers()
```

File: discord/client.py [hunk 2]
```diff
@@ -246,12 +246,18 @@ def __init__(
...
 
         if VoiceClient.warn_nacl:
             VoiceClient.warn_nacl = False
             _log.warning("PyNaCl is not installed, voice will NOT be supported")
 
     # internals
+    def _prepeare_handlers(self):
```
Comment (VincentRPS) [inline discord/client.py:257]:
```suggestion
    def _prepare_handlers(self):
```

File: discord/client.py [hunk 3]
```diff
@@ -1133,34 +1136,68 @@ def _check(*args):
 
     # event registration
 
-    def event(self, coro: Coro) -> Coro:
+    def event(self, event: str) -> Coro:
```
Comment (VincentRPS) [inline discord/client.py:1139]: sorry if I'm misunderstanding, but wouldn't this be a breaking change? Maybe keep how it currently is and add support on `@listen` as well

File: discord/client.py [hunk 4]
```diff
@@ -1133,34 +1136,68 @@ def _check(*args):
...
             async def on_ready():
                 print('Ready!')
         """
 
-        if not asyncio.iscoroutinefunction(coro):
-            raise TypeError("event registered must be a coroutine function")
+        def _event(coro: Coro) -> Coro:
+            """Decorator to register the event."""
+            if not asyncio.iscoroutinefunction(coro):
+                raise TypeError("event registered must be a coroutine function")
```
Comment (VincentRPS) [inline discord/client.py:1176]:
```suggestion
                raise TypeError("Event registered must be a coroutine function")
```
capitalization consistency

File: discord/client.py [hunk 5]
```diff
@@ -1133,34 +1136,68 @@ def _check(*args):
...
         Example
         -------
 
         .. code-block:: python3
 
+            @client.event("on_ready")
+            async def ready_handler():
+                print('Ready!')
+
             @client.event
```
Comment (VincentRPS) [inline discord/client.py:1168]: have you tested whether this method works?

**Review by VincentRPS — CHANGES_REQUESTED** [review]

File: discord/client.py (hunk 3 above)
Comment (davidhozic) [reply discord/client.py:1139]: This is used to be a decorator, meaning that it was used like this:
```py
import discord

client = discord.Client()

@client.event
async def on_ready():
    pass
```

Meaning that users could not use the argument as a keyword argument

File: discord/client.py (hunk 5 above)
Comment (davidhozic) [reply discord/client.py:1168]: Yes

File: discord/client.py (hunk 3 above)
Comment (davidhozic) [reply discord/client.py:1139]: Doesn't ``@listen`` already have similar functionality? It allows you to pass an event name as parameter or no parameters , but still with brackets, for it to use the handler name.

Comment (davidhozic) [reply discord/client.py]: Is it ok like this?

**Comment by davidhozic** [comment]
Above change was from a master branch rebase

**Review by Lulalaby — APPROVED** [review]

**Review by VincentRPS — APPROVED** [review]

**Pull request merged**